pkg/ccl/importccl/import_stmt.go

```diff
@@ -472,39 +471,80 @@ func importPlanHook(
+				if err := backupccl.CheckTableExists(ctx, txn, parentID, tableDesc.Name); err != nil {
+					return err
+				}
 			}
-		}
-		// Verification steps have passed, generate a new table ID if we're
-		// restoring. We do this last because we want to avoid calling
-		// GenerateUniqueDescID if there's any kind of error above.
-		// Reserving a table ID now means we can avoid the rekey work during restore.
-		tableRewrites := make(backupccl.TableRewriteMap)
-		newSeqVals := make(map[sqlbase.ID]int64, len(seqVals))
-		for _, tableDesc := range tableDescs {
-			id, err := sql.GenerateUniqueDescID(ctx, p.ExecCfg().DB)
-			if err != nil {
+			// Verification steps have passed, generate a new table ID if we're
+			// restoring. We do this last because we want to avoid calling
+			// GenerateUniqueDescID if there's any kind of error above.
+			// Reserving a table ID now means we can avoid the rekey work during restore.
+			tableRewrites := make(backupccl.TableRewriteMap)
+			newSeqVals := make(map[sqlbase.ID]int64, len(seqVals))
+			for _, tableDesc := range tableDescs {
+				id, err := sql.GenerateUniqueDescID(ctx, p.ExecCfg().DB)
+				if err != nil {
+					return err
+				}
+				tableRewrites[tableDesc.ID] = &jobspb.RestoreDetails_TableRewrite{
+					TableID:  id,
+					ParentID: parentID,
+				}
+				if v, ok := seqVals[tableDesc.ID]; ok {
+					newSeqVals[id] = v
+				}
+			}
+			seqVals = newSeqVals
+			if err := backupccl.RewriteTableDescs(tableDescs, tableRewrites, ""); err != nil {
 				return err
 			}
-			tableRewrites[tableDesc.ID] = &jobspb.RestoreDetails_TableRewrite{
-				TableID:  id,
-				ParentID: parentID,
+
+			for i := range tableDescs {
+				tableDescs[i].State = sqlbase.TableDescriptor_IMPORTING
 			}
-			newSeqVals[id] = seqVals[tableDesc.ID]
-		}
-		seqVals = newSeqVals
-		if err := backupccl.RewriteTableDescs(tableDescs, tableRewrites, ""); err != nil {
+
+			seqValKVs := make([]roachpb.KeyValue, 0, len(seqVals))
+			for i := range tableDescs {
+				if v, ok := seqVals[tableDescs[i].ID]; ok && v != 0 {
+					key, val, err := sql.MakeSequenceKeyVal(tableDescs[i], v, false)
+					if err != nil {
+						return err
+					}
+					kv := roachpb.KeyValue{Key: key}
+					kv.Value.SetInt(val)
+					seqValKVs = append(seqValKVs, kv)
+				}
+			}
+
+			// Write the new TableDescriptors and flip the namespace entries over to
+			// them. After this call, any queries on a table will be served by the newly
+			// imported data.
+			if err := backupccl.WriteTableDescs(ctx, txn, nil, tableDescs, p.User(), p.ExecCfg().Settings, seqValKVs); err != nil {
+				return pgerror.Wrapf(err, pgerror.CodeDataExceptionError, "creating tables")
+			}
+
+			for i := range tableDescs {
+
+				log.Infof(ctx, "staging tbl via cput %v \nnil\n -> \n%v\n", sqlbase.MakeDescMetadataKey(tableDescs[i].ID), sqlbase.WrapDescriptor(tableDescs[i]))
+
+			}
+			// TODO(dt): we should be creating the job with this txn too. Once a job
```

Comment:
how hard is this? seems kinda important.

Reply:
agreed it is important. At first glance, it'll be pretty tricky and will require refactoring the jobs code a bit. I looked at it for a bit but decided that it was enough going to involve enough moving things around that I'd rather not do it here and leave it to a self-contained change later, but I'd certainly consider something that would need to be addressed before release.